howtos/diffs/howto-ensembling.diff

```diff
@@ -0,0 +1,142 @@
+diff --git a/examples/mnist/train.py b/examples/mnist/train.py
+index e93321f..3d2e0b5 100644
+--- a/examples/mnist/train.py
++++ b/examples/mnist/train.py
+@@ -18,6 +18,7 @@ This script trains a simple Convolutional Neural Net on the MNIST dataset.
+ The data is loaded using tensorflow_datasets.
+ 
+ """
++import functools
+ 
+ from absl import app
+ from absl import flags
+@@ -85,9 +86,11 @@ def create_model(key):
+   return model
+ 
+ 
+-def create_optimizer(model, learning_rate, beta):
+-  optimizer_def = optim.Momentum(learning_rate=learning_rate, beta=beta)
+-  optimizer = optimizer_def.create(model)
++@jax.pmap
++def create_optimizers(rng):
++  optimizer_def = optim.Momentum(
++      learning_rate=FLAGS.learning_rate, beta=FLAGS.momentum)
++  optimizer = optimizer_def.create(create_model(rng))
+   return optimizer
+ 
+ 
+@@ -110,7 +113,7 @@ def compute_metrics(logits, labels):
+   return metrics
+ 
+ 
+-@jax.jit
++@functools.partial(jax.pmap)
+ def train_step(optimizer, batch):
+   """Train for a single step."""
+   def loss_fn(model):
+@@ -122,13 +125,17 @@ def train_step(optimizer, batch):
+   return optimizer, metrics
+ 
+ 
+-@jax.jit
++@jax.pmap
+ def eval_step(model, batch):
+   logits = model(batch['image'])
+   return compute_metrics(logits, batch['label'])
+ 
+ 
+-def train_epoch(optimizer, train_ds, batch_size, epoch, rng):
++def replicate(tree_obj):
```

Comment:
I think @jheek suggested you replace this with `jax_utils.replicate`?

Reply:
Yes, I've now updated it. I've also tested the code in Colab. Note this is a bit cumbersome, since it requires multiple devices, so I ended up changing the google internal version and used Colab internal.